src/server/page_server.cpp

```diff
@@ -371,6 +377,79 @@ page_server::async_disconnect_handler::disconnect_loop ()
     }
 }
 
+void page_server::pts_mvcc_tracker::init_oldest_active_mvccid (const std::string &pts_channel_id)
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+  /*
+   * The entry must not already be present. If the same passive transaction server has been connected
+   * before, the entry must have been removed when the PTS disconnected or when the connection
+   *  to the PTS was aborted.
+   */
+  assert (m_pts_oldest_active_mvccids.find (pts_channel_id) == m_pts_oldest_active_mvccids.end());
+
+  /*
+   * MVCCID_ALL_VISIBLE means that it hasn't yet received. It will prevent the ATS to run vacuum.
+   * This is a guard for the window in which a PTS is connected but has't sent its oldest active mvccid.
+   * In this window, if we vaccum without considering the PTS, we possibly end up cleaning up the data
+   * a read-only transaction on the PTS see.
+   */
+  m_pts_oldest_active_mvccids[pts_channel_id] = MVCCID_ALL_VISIBLE;
+}
+
+void page_server::pts_mvcc_tracker::update_oldest_active_mvccid (const std::string &pts_channel_id, const MVCCID mvccid)
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+
+  /*
+   * 1. The entry is already created when ths PTS is connected.
+   * 2. It is updated by the PTS only when it move foward.
+   *    Without update, it is MVCCID_ALL_VISIBLE by default, which is lower than any mvccid assigned.
+   */
+  assert (m_pts_oldest_active_mvccids.find (pts_channel_id) != m_pts_oldest_active_mvccids.end());
+  assert (m_pts_oldest_active_mvccids[pts_channel_id] < mvccid);
```

Comment:
```suggestion
  assert (m_pts_oldest_active_mvccids[pts_channel_id] <= mvccid);
```

Reply:
I think `assert (m_pts_oldest_active_mvccids[pts_channel_id] < mvccid);` is correct.

Reply:
I think that a PTS can send same oldest mvccid to PS. But It cannot. (in http://jira.cubrid.org/browse/LETS-547)

Your condition is correct.